[WebGPU] iOS validation layer rejects drawIndexed calls with zero indices

mwyrzykowski opened the pull request:
#### 18401156b3ed582986e32520554043c0972755d5
<pre>
[WebGPU] iOS validation layer rejects drawIndexed calls with zero indices
<a href="https://bugs.webkit.org/show_bug.cgi?id=290978">https://bugs.webkit.org/show_bug.cgi?id=290978</a>
<a href="https://rdar.apple.com/148379856">rdar://148379856</a>

Reviewed by Cameron McCormack.

drawIndexedIndirect with an indirect buffer which looks like
{
  .indexCount = 0,
  .instanceCount = 12,
  .indexStart = 0,
  .baseVertex = 9,
  .baseInstance = 0
}

reported out of bounds vertex fetches, even though the index count was zero.

* LayoutTests/fast/webgpu/nocrash/fuzz-290978-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-290978.html: Added.
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::indexBufferClampPipeline):

Canonical link: <a href="https://commits.webkit.org/293251@main">https://commits.webkit.org/293251@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e8a70ce7d0104f73dd14b5211b3d8b5cff2a3be

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98062 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17693 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7920 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103177 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48591 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26144 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74657 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31843 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101066 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13622 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88605 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55017 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6538 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48033 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6620 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105555 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25148 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18312 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83644 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25521 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84786 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83098 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21045 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27737 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5415 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25107 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30281 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24927 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28243 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26502 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam reviewed (approved):
Is there a CTS test that covers the validity of indexed draws when the count is zero but the offset is out of bounds?

mwyrzykowski committed:
[WebGPU] iOS validation layer rejects drawIndexed calls with zero indices
https://bugs.webkit.org/show_bug.cgi?id=290978
rdar://148379856

Reviewed by Cameron McCormack.

drawIndexedIndirect with an indirect buffer which looks like
{
  .indexCount = 0,
  .instanceCount = 12,
  .indexStart = 0,
  .baseVertex = 9,
  .baseInstance = 0
}

reported out of bounds vertex fetches, even though the index count was zero.

* LayoutTests/fast/webgpu/nocrash/fuzz-290978-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-290978.html: Added.
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::indexBufferClampPipeline):

Canonical link: https://commits.webkit.org/293251@main
